lib/matplotlib/text.py

```diff
@@ -66,24 +66,19 @@ def _get_text_metrics_with_cache(renderer, text, fontprop, ismath, dpi):
     """Call ``renderer.get_text_width_height_descent``, caching the results."""
     # Cached based on a copy of fontprop so that later in-place mutations of
     # the passed-in argument do not mess up the cache.
-    func = _get_text_metrics_function(weakref.ref(renderer))
+    func = _get_text_metrics_function(renderer)
     return func(text, fontprop.copy(), ismath, dpi)
 
 
-@functools.lru_cache(25)
-def _get_text_metrics_function(renderer_ref):
-
-    @functools.lru_cache(4096)
-    def _text_metrics(text, fontprop, ismath, dpi):
-        # dpi is unused, but participates in cache invalidation (via the renderer).
-        renderer = renderer_ref()
-        if renderer is None:
-            raise RuntimeError(
-                "You are trying to get the font metrics for a renderer that no "
-                "longer exists.  There is likely a deeper bug someplace else "
-                "as this should never happen."
-            )
-        return renderer.get_text_width_height_descent(text, fontprop, ismath)
+# use mutable default to carry hidden global state
+def _get_text_metrics_function(renderer, _cache=weakref.WeakKeyDictionary()):
+    if (_text_metrics := _cache.get(renderer, None)) is None:
```

Comment:
It seems something is missing here: If the renderer is cached this if branch is not executed, i.e. the whole function does nothing and returns `None`?!?